splash-reward-distributor/src/engine/queue.rs

```diff
@@ -118,17 +134,59 @@ where
                     }
                     QueueCmd::Update(id, task) => update(&tx, &tables, id, task),
                     QueueCmd::Cancel(id) => cancel(&tx, &tables, id),
-                    QueueCmd::Done(id) => done(&tx, &tables, id),
+                    QueueCmd::Done(id, tx_hash) => {
+                        done(&tx, &tables, id);
+                        done_pairs.push((id, tx_hash));
+                    }
                     QueueCmd::AdvanceClocks(time) => advance_clocks(&tx, &tables, time),
                     QueueCmd::DowngradeClocks(time) => downgrade_clocks(&tx, &tables, time),
                 }
             }
+
+            // If there's at least one QueueCmd::Done(..), need to associate it with the TX hash
+            if let Some(first_tx_hash) = done_pairs.first().map(|(_, tx_hash)| *tx_hash) {
+                let task_ids: Vec<_> = done_pairs
+                    .iter()
+                    .map(|(task_id, tx_hash)| {
+                        assert_eq!(*tx_hash, first_tx_hash);
+                        *task_id
+                    })
+                    .collect();
+
+                let mapped_tasks: Option<Vec<TaskId>> = get_task_ids(&tx, &tables, first_tx_hash);
+
+                if let Some(mapped_tasks) = mapped_tasks {
+                    // We would enter this region if a TX is confirmed in a block.
+                    assert_eq!(mapped_tasks, task_ids);
+                } else {
+                    insert_task_ids(&tx, &tables, first_tx_hash, task_ids);
+                }
+            }
```

Comment:
There's a way to simplify this logic: in the new table store mapping as `key{tx_id:task_id} -> value{empty}`, you'll be able to retrieve all tasks by tx_id using it as a key prefix `key{tx_id:}`. Plus simply update done() method to  insert `key{tx_id:task_id} -> value{empty}` entry. 
In the current version you have to have all these consistency checks and extra update logic bc you store all tasks under one key.

Reply:
Done